File: apple/Sources/ReactNativeWatchOS/Modifiers/Rotation3DEffect.swift
import SwiftUI

/// SwiftUI `.rotation3DEffect(_:axis:anchor:)`. Rotates the view about the
/// 3D axis `(x, y, z)` by `degrees`. Axis defaults to the Y axis `(0,1,0)`
/// when none of the components are supplied. `anchorZ` / `perspective` are
/// not bridged and use SwiftUI's defaults.
enum RNWRotation3DEffectModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("rotation3DEffect") { view, params, _ in
            let degrees = params.double("degrees") ?? 0
            let x = params.cgFloat("axisX")
            let y = params.cgFloat("axisY")
            let z = params.cgFloat("axisZ")
            // Default to the Y axis only when no component is provided.
            let axis: (x: CGFloat, y: CGFloat, z: CGFloat)
            if x == nil, y == nil, z == nil {
                axis = (0, 1, 0)
            } else {
                axis = (x ?? 0, y ?? 0, z ?? 0)
            }
            let anchor = RNWTransformAnchor.unitPoint(params.string("anchor"))
            return AnyView(view.rotation3DEffect(
                .degrees(degrees),
                axis: axis,
                anchor: anchor
            ))
        }
    }
}

File: src/modifiers/rotation3DEffect.ts
import { createModifier } from './createModifier';
import type { TransformAnchor } from './rotationEffect';

export interface Rotation3DEffectParams {
  /// Rotation angle in degrees.
  degrees: number;
  /// X component of the rotation axis. Defaults to `0`.
  axisX?: number;
  /// Y component of the rotation axis. Defaults to `1` (rotate about Y).
  axisY?: number;
  /// Z component of the rotation axis. Defaults to `0`.
  axisZ?: number;
  /// 2D anchor point for the rotation. Defaults to `'center'`.
  anchor?: TransformAnchor;
}

/// SwiftUI `.rotation3DEffect(_:axis:anchor:)`. Rotates the view about the
/// 3D axis `(axisX, axisY, axisZ)` by `degrees`. The optional `anchorZ`
/// and perspective parameters are not bridged; defaults are used.
export function rotation3DEffect(params: Rotation3DEffectParams) {
  return createModifier('rotation3DEffect', params);
}
